File: prototype-python/fuji_xs20/recipe_parser.py
"""Parse text recipe kiểu Fuji X Weekly → dict recipe chuẩn hoá.

Bản tham chiếu Python (test được cục bộ) để chốt logic trước khi port sang Swift
(RecipeParser.swift). Trả về dict với tên enum khớp FujiKit + số + notes.

Chịu được các ca thực tế: tên recipe đứng trước tên film sim, film sim xuống dòng
("Classic"/"Negative"), decimal kiểu châu Âu ("+2,5"), lỗi OCR ("t1 Red"=+1), nhiễu
ký tự lạ ("の"), header ghép ("NOISE REDUCTION/HIGH ISO NR").
"""

from __future__ import annotations

import re
import unicodedata

# --- Header: (các alias CHÍNH XÁC sau chuẩn hoá, field key). Thứ tự: cụ thể → chung.
_HEADERS: list[tuple[list[str], str]] = [
    (["FILM SIMULATION", "FILM SIM"], "film"),
    (["DYNAMIC RANGE", "D RANGE", "DR"], "dr"),
    (["GRAIN EFFECT", "GRAIN"], "grain"),
    (["COLOR CHROME EFFECT BLUE", "COLOR CHROME FX BLUE", "COLOR CHROME BLUE",
      "COLOUR CHROME FX BLUE"], "cc_blue"),
    (["COLOR CHROME EFFECT", "COLOR CHROME FX", "COLOR CHROME", "COLOUR CHROME"], "cc_effect"),
    (["WHITE BALANCE", "WB"], "wb"),
    (["HIGHLIGHT TONE", "HIGHLIGHT", "HIGH LIGHT"], "highlight"),
    (["SHADOW TONE", "SHADOW"], "shadow"),
    (["SHARPNESS", "SHARPENING"], "sharpness"),
    (["NOISE REDUCTION/HIGH ISO NR", "NOISE REDUCTION", "HIGH ISO NR", "NR"], "nr"),
    (["CLARITY"], "clarity"),
    (["COLOR", "COLOUR"], "color"),
    (["ISO"], "iso"),
    (["EXPOSURE COMPENSATION", "EXP COMPENSATION", "EXP. COMPENSATION", "EXP COMP", "EV"], "exp_comp"),
]

# --- Film simulation aliases → tên enum FujiKit. Dài trước (greedy đúng).
_FILM: list[tuple[str, str]] = [
    ("ETERNA BLEACH BYPASS", "eternaBleachBypass"),
    ("BLEACH BYPASS", "eternaBleachBypass"),
    ("ETERNA CINEMA", "eterna"), ("ETERNA", "eterna"),
    ("CLASSIC NEGATIVE", "classicNeg"), ("CLASSIC NEG", "classicNeg"),
    ("NOSTALGIC NEGATIVE", "nostalgicNeg"), ("NOSTALGIC NEG", "nostalgicNeg"),
    ("CLASSIC CHROME", "classicChrome"),
    ("PRO NEG HI", "proNegHi"), ("PRO NEG. HI", "proNegHi"), ("PRONEG HI", "proNegHi"),
    ("PRO NEG STD", "proNegStd"), ("PRO NEG. STD", "proNegStd"),
    ("REALA ACE", "realaAce"), ("REALA", "realaAce"),
    ("ACROS+YE", "acrosYe"), ("ACROS+R", "acrosR"), ("ACROS+G", "acrosG"), ("ACROS", "acros"),
    ("MONOCHROME+YE", "monochromeYe"), ("MONOCHROME+R", "monochromeR"),
    ("MONOCHROME+G", "monochromeG"), ("MONOCHROME", "monochrome"),
    ("PROVIA/STANDARD", "provia"), ("PROVIA", "provia"), ("STANDARD", "provia"),
    ("VELVIA/VIVID", "velvia"), ("VELVIA", "velvia"),
    ("ASTIA/SOFT", "astia"), ("ASTIA", "astia"),
    ("SEPIA", "sepia"),
]

_WB_MODE: list[tuple[str, str]] = [
    ("COLOR TEMPERATURE", "colorTemp"), ("COLOUR TEMPERATURE", "colorTemp"), ("KELVIN", "colorTemp"),
    ("AUTO WHITE PRIORITY", "autoWhitePriority"), ("WHITE PRIORITY", "autoWhitePriority"),
    ("AUTO AMBIENCE", "autoAmbiencePriority"), ("AMBIENCE", "autoAmbiencePriority"),
    ("DAYLIGHT", "daylight"), ("SHADE", "shade"),
    ("INCANDESCENT", "incandescent"), ("UNDERWATER", "underwater"),
    ("FLUORESCENT 1", "fluorescent1"), ("FLUORESCENT 2", "fluorescent2"),
    ("FLUORESCENT 3", "fluorescent3"),
    ("AUTO", "auto"),
]

_NR_MAP = {-4: "m4", -3: "m3", -2: "m2", -1: "m1", 0: "std", 1: "p1", 2: "p2", 3: "p3", 4: "p4"}


def _norm(s: str) -> str:
    s = unicodedata.normalize("NFKC", s).upper().strip()
    s = re.sub(r"[：:]+$", "", s).strip()          # bỏ ':' cuối
    s = re.sub(r"\s+", " ", s)
    return s


def _match_header(line: str) -> str | None:
    n = _norm(line)
    for aliases, key in _HEADERS:
        if n in aliases:
            return key
    return None


def _num(s: str):
    """Số đầu tiên trong s, hỗ trợ dấu và decimal ',' hoặc '.'. None nếu không có."""
    m = re.search(r"[+-]?\d+(?:[.,]\d+)?", s)
    if not m:
        return None
    v = float(m.group().replace(",", "."))
    return int(v) if v == int(v) else v


def _clean_name(s: str) -> str:
    # bỏ ký tự không phải chữ Latin/số/space (nhiễu OCR như 'の'), gọn khoảng trắng
    s = "".join(ch for ch in s if ch.isascii() or ch.isspace())
    return re.sub(r"\s+", " ", s).strip(" -–—:·")


def parse(text: str) -> dict:
    # 1) tách section theo header; lines trước header đầu = preamble (tên recipe kiểu tiêu đề)
    sections: dict[str, list[str]] = {}
    preamble: list[str] = []
    current: str | None = None
    for raw in text.splitlines():
        line = raw.strip()
        if not line:
            continue
        key = _match_header(line)
        if key:
            current = key
            sections.setdefault(key, [])
        elif current:
            sections[current].append(line)
        else:
            preamble.append(line)

    out: dict = {"name": None, "notes": {}}

    # 2) FILM SIMULATION + tên recipe (giữ nguyên hoa/thường của tên)
    if "film" in sections:
        orig = " ".join(sections["film"])
        norm = _norm(orig)
        film, alias_found, idx = None, None, -1
        for alias, enum in _FILM:
            p = norm.rfind(alias)           # ưu tiên xuất hiện CUỐI (film sim đứng sau tên)
            if p != -1 and p >= idx:
                film, alias_found, idx = enum, alias, p
        out["film_simulation"] = film
        if alias_found:
            pat = re.compile(r"\s+".join(re.escape(w) for w in alias_found.split(" ")), re.I)
            matches = list(pat.finditer(orig))
            if matches:
                out["name"] = _clean_name(orig[:matches[-1].start()]) or None

    # tên kiểu tiêu đề đứng đầu (nếu section film không cho tên)
    if not out["name"] and preamble:
        out["name"] = _clean_name(" ".join(preamble)) or None

    # 3) DR
    if "dr" in sections:
        t = _norm(" ".join(sections["dr"]))
        if "AUTO" in t:
            out["dynamic_range"] = "auto"
        else:
            n = _num(t)
            out["dynamic_range"] = {100: "dr100", 200: "dr200", 400: "dr400"}.get(int(n)) if n else None

    # 4) GRAIN
    if "grain" in sections:
        t = _norm(" ".join(sections["grain"]))
        if "OFF" in t:
            out["grain"] = "offLarge" if "LARGE" in t else "off"
        else:
            strong = "STRONG" in t
            large = "LARGE" in t
            out["grain"] = {(False, False): "weakSmall", (True, False): "strongSmall",
                            (False, True): "weakLarge", (True, True): "strongLarge"}[(strong, large)]

    # 5) COLOR CHROME EFFECT / BLUE
    def _chrome(t: str) -> str:
        t = _norm(t)
        if "STRONG" in t: return "strong"
        if "WEAK" in t: return "weak"
        return "off"
    if "cc_effect" in sections:
        out["color_chrome_effect"] = _chrome(" ".join(sections["cc_effect"]))
    if "cc_blue" in sections:
        out["color_chrome_blue"] = _chrome(" ".join(sections["cc_blue"]))

    # 6) WB: mode + kelvin + shift R/B
    if "wb" in sections:
        t = _norm(" ".join(sections["wb"]))
        km = re.search(r"(\d{4,5})\s*K", t)
        if km:
            out["wb_kelvin"] = int(km.group(1))
            out["white_balance"] = "colorTemp"
        else:
            for alias, enum in _WB_MODE:
                if alias in t:
                    out["white_balance"] = enum
                    break
        mr = re.search(r"([+-]?\d+)\s*RED", t) or re.search(r"RED\s*([+-]?\d+)", t)
        mb = re.search(r"([+-]?\d+)\s*BLUE", t) or re.search(r"BLUE\s*([+-]?\d+)", t)
        if mr: out["wb_shift_red"] = int(mr.group(1))
        if mb: out["wb_shift_blue"] = int(mb.group(1))

    # 7) tones / color / sharp / clarity
    for key, field in [("highlight", "highlight_tone"), ("shadow", "shadow_tone")]:
        if key in sections:
            out[field] = _num(" ".join(sections[key]))
    for key, field in [("color", "color"), ("sharpness", "sharpness"), ("clarity", "clarity")]:
        if key in sections:
            n = _num(" ".join(sections[key]))
            out[field] = int(n) if n is not None else None

    # 8) NR số → enum
    if "nr" in sections:
        n = _num(" ".join(sections["nr"]))
        if n is not None:
            out["noise_reduction"] = _NR_MAP.get(int(n))

    # 9) ISO / EXP COMP → notes
    if "iso" in sections:
        out["notes"]["iso"] = " ".join(sections["iso"]).strip()
    if "exp_comp" in sections:
        out["notes"]["exposure_compensation"] = " ".join(sections["exp_comp"]).strip()

    return out

File: swift/FujiKit/Sources/FujiKit/RecipeParser.swift
import Foundation

/// Parse text recipe kiểu Fuji X Weekly → Recipe. Port trung thực từ
/// prototype-python/fuji_xs20/recipe_parser.py (đã test 2 ca). Chịu được: tên recipe
/// trước film sim hoặc ở tiêu đề, film sim xuống dòng, decimal "+2,5", OCR "t1 Red"=+1,
/// nhiễu ký tự lạ, header ghép "NOISE REDUCTION/HIGH ISO NR".
public enum RecipeParser {

    // (alias CHÍNH XÁC sau chuẩn hoá, field). Thứ tự: cụ thể → chung.
    private static let headers: [([String], String)] = [
        (["FILM SIMULATION", "FILM SIM"], "film"),
        (["DYNAMIC RANGE", "D RANGE", "DR"], "dr"),
        (["GRAIN EFFECT", "GRAIN"], "grain"),
        (["COLOR CHROME EFFECT BLUE", "COLOR CHROME FX BLUE", "COLOR CHROME BLUE", "COLOUR CHROME FX BLUE"], "cc_blue"),
        (["COLOR CHROME EFFECT", "COLOR CHROME FX", "COLOR CHROME", "COLOUR CHROME"], "cc_effect"),
        (["WHITE BALANCE", "WB"], "wb"),
        (["HIGHLIGHT TONE", "HIGHLIGHT", "HIGH LIGHT"], "highlight"),
        (["SHADOW TONE", "SHADOW"], "shadow"),
        (["SHARPNESS", "SHARPENING"], "sharpness"),
        (["NOISE REDUCTION/HIGH ISO NR", "NOISE REDUCTION", "HIGH ISO NR", "NR"], "nr"),
        (["CLARITY"], "clarity"),
        (["COLOR", "COLOUR"], "color"),
        (["ISO"], "iso"),
        (["EXPOSURE COMPENSATION", "EXP COMPENSATION", "EXP. COMPENSATION", "EXP COMP", "EV"], "exp_comp"),
    ]

    private static let films: [(String, FilmSimulation)] = [
        ("ETERNA BLEACH BYPASS", .eternaBleachBypass), ("BLEACH BYPASS", .eternaBleachBypass),
        ("ETERNA CINEMA", .eterna), ("ETERNA", .eterna),
        ("CLASSIC NEGATIVE", .classicNeg), ("CLASSIC NEG", .classicNeg),
        ("NOSTALGIC NEGATIVE", .nostalgicNeg), ("NOSTALGIC NEG", .nostalgicNeg),
        ("CLASSIC CHROME", .classicChrome),
        ("PRO NEG HI", .proNegHi), ("PRO NEG. HI", .proNegHi), ("PRONEG HI", .proNegHi),
        ("PRO NEG STD", .proNegStd), ("PRO NEG. STD", .proNegStd),
        ("REALA ACE", .realaAce), ("REALA", .realaAce),
        ("ACROS+YE", .acrosYe), ("ACROS+R", .acrosR), ("ACROS+G", .acrosG), ("ACROS", .acros),
        ("MONOCHROME+YE", .monochromeYe), ("MONOCHROME+R", .monochromeR),
        ("MONOCHROME+G", .monochromeG), ("MONOCHROME", .monochrome),
        ("PROVIA/STANDARD", .provia), ("PROVIA", .provia), ("STANDARD", .provia),
        ("VELVIA/VIVID", .velvia), ("VELVIA", .velvia),
        ("ASTIA/SOFT", .astia), ("ASTIA", .astia),
        ("SEPIA", .sepia),
    ]

    private static let wbModes: [(String, WhiteBalance)] = [
        ("COLOR TEMPERATURE", .colorTemp), ("COLOUR TEMPERATURE", .colorTemp), ("KELVIN", .colorTemp),
        ("AUTO WHITE PRIORITY", .autoWhitePriority), ("WHITE PRIORITY", .autoWhitePriority),
        ("AUTO AMBIENCE", .autoAmbiencePriority), ("AMBIENCE", .autoAmbiencePriority),
        ("DAYLIGHT", .daylight), ("SHADE", .shade),
        ("INCANDESCENT", .incandescent), ("UNDERWATER", .underwater),
        ("FLUORESCENT 1", .fluorescent1), ("FLUORESCENT 2", .fluorescent2), ("FLUORESCENT 3", .fluorescent3),
        ("AUTO", .auto),
    ]

    private static let nrMap: [Int: NoiseReduction] = [
        -4: .m4, -3: .m3, -2: .m2, -1: .m1, 0: .std, 1: .p1, 2: .p2, 3: .p3, 4: .p4,
    ]

    // ---- helpers ----
    private static func norm(_ s: String) -> String {
        var t = s.precomposedStringWithCanonicalMapping.uppercased()
        t = t.trimmingCharacters(in: .whitespacesAndNewlines)
        while t.hasSuffix(":") || t.hasSuffix("：") { t.removeLast() }
        t = t.replacingOccurrences(of: "\\s+", with: " ", options: .regularExpression)
        return t.trimmingCharacters(in: .whitespaces)
    }

    private static func number(_ s: String) -> Double? {
        guard let m = firstMatch("[+-]?\\d+(?:[.,]\\d+)?", s) else { return nil }
        return Double(m.replacingOccurrences(of: ",", with: "."))
    }

    private static func firstMatch(_ pattern: String, _ s: String, group: Int = 0,
                                   caseInsensitive: Bool = false) -> String? {
        let opts: NSRegularExpression.Options = caseInsensitive ? [.caseInsensitive] : []
        guard let re = try? NSRegularExpression(pattern: pattern, options: opts) else { return nil }
        let ns = NSRange(s.startIndex..., in: s)
        guard let m = re.firstMatch(in: s, range: ns), m.numberOfRanges > group,
              let r = Range(m.range(at: group), in: s) else { return nil }
        return String(s[r])
    }

    private static func matchHeader(_ line: String) -> String? {
        let n = norm(line)
        for (aliases, key) in headers where aliases.contains(n) { return key }
        return nil
    }

    /// Alias phẳng, sắp DÀI NHẤT trước (để "COLOR CHROME EFFECT BLUE" không bị "COLOR CHROME
    /// EFFECT" nuốt khi so tiền tố).
    private static let aliasesLongestFirst: [String] =
        headers.flatMap { $0.0 }.sorted { $0.count > $1.count }

    /// Dòng OCR kiểu CỘT "FILM SIMULATION Reala Ace" (nhãn + giá trị cùng dòng, KHÔNG dấu ':')
    /// → (header, value). Trả nil nếu dòng không mở đầu bằng header đã biết + theo sau là giá trị
    /// (vd "DR400" không có khoảng trắng → nil; "FILM SIMULATION" trơ trọi → nil để giữ format 2 dòng).
    public static func splitLeadingHeader(_ line: String) -> (header: String, value: String)? {
        let trimmed = line.trimmingCharacters(in: .whitespaces)
        let n = norm(trimmed)
        for alias in aliasesLongestFirst {
            guard n != alias, n.hasPrefix(alias + " ") else { continue }
            let words = trimmed.split(separator: " ", omittingEmptySubsequences: true)
            let headWords = alias.split(separator: " ").count
            guard words.count > headWords else { return nil }
            let value = words.dropFirst(headWords).joined(separator: " ")
            guard !value.isEmpty else { return nil }
            return (words.prefix(headWords).joined(separator: " "), value)
        }
        return nil
    }

    private static func cleanName(_ s: String) -> String {
        let ascii = String(s.unicodeScalars.filter { $0.isASCII || $0 == " " })
        return ascii.replacingOccurrences(of: "\\s+", with: " ", options: .regularExpression)
            .trimmingCharacters(in: CharacterSet(charactersIn: " -–—:·"))
    }

    /// Tìm film sim trong 1 chuỗi (ưu tiên xuất hiện cuối). Trả (film, alias).
    private static func findFilm(_ text: String) -> (FilmSimulation, String)? {
        let n = norm(text)
        var best = -1
        var result: (FilmSimulation, String)?
        for (alias, film) in films {
            if let rng = n.range(of: alias, options: .backwards) {
                let pos = n.distance(from: n.startIndex, to: rng.lowerBound)
                if pos >= best { best = pos; result = (film, alias) }
            }
        }
        return result
    }

    /// Vị trí bắt đầu của lần khớp CUỐI của alias (whitespace linh hoạt) trong orig.
    private static func lastAliasStart(_ alias: String, in orig: String) -> String.Index? {
        let pat = alias.split(separator: " ").map { NSRegularExpression.escapedPattern(for: String($0)) }.joined(separator: "\\s+")
        guard let re = try? NSRegularExpression(pattern: pat, options: [.caseInsensitive]) else { return nil }
        let ns = NSRange(orig.startIndex..., in: orig)
        guard let last = re.matches(in: orig, range: ns).last, let rr = Range(last.range, in: orig) else { return nil }
        return rr.lowerBound
    }

    // ---- main ----
    public static func parse(_ text: String) -> Recipe {
        var sections: [String: [String]] = [:]
        var preamble: [String] = []
        var current: String?
        for raw in text.split(whereSeparator: \.isNewline) {
            let line = raw.trimmingCharacters(in: .whitespaces)
            if line.isEmpty { continue }
            // "Header: value" cùng dòng
            if let colon = line.firstIndex(of: ":") {
                let left = String(line[..<colon])
                let right = String(line[line.index(after: colon)...]).trimmingCharacters(in: .whitespaces)
                if let k = matchHeader(left) {
                    if sections[k] == nil { sections[k] = [] }
                    if !right.isEmpty { sections[k]?.append(right) }
                    current = k
                    continue
                }
            }
            if let key = matchHeader(line) { current = key; sections[key] = sections[key] ?? [] }
            else if let c = current { sections[c, default: []].append(line) }
            else { preamble.append(line) }
        }

        var r = Recipe()

        // FILM + tên: thử section (format cũ), rồi preamble (format mới: tên + film sim đứng riêng)
        if let lines = sections["film"], let (film, alias) = findFilm(lines.joined(separator: " ")) {
            r.filmSimulation = film
            let orig = lines.joined(separator: " ")
            if let start = lastAliasStart(alias, in: orig) {
                let nm = cleanName(String(orig[orig.startIndex..<start]))
                if !nm.isEmpty { r.name = nm }
            }
        }
        if r.filmSimulation == nil, !preamble.isEmpty,
           let (film, alias) = findFilm(preamble.joined(separator: " ")) {
            r.filmSimulation = film
            let words = alias.split(separator: " ").map { NSRegularExpression.escapedPattern(for: String($0)) }.joined(separator: "\\s+")
            let nameLines = preamble.filter { firstMatch(words, $0, caseInsensitive: true) == nil }
            let nm = cleanName(nameLines.joined(separator: " "))
            if !nm.isEmpty { r.name = nm }
        }
        if r.name == "Untitled", !preamble.isEmpty {
            let nm = cleanName(preamble.joined(separator: " "))
            if !nm.isEmpty { r.name = nm }
        }

        // DR
        if let l = sections["dr"] {
            let t = norm(l.joined(separator: " "))
            if t.contains("AUTO") { r.dynamicRange = .auto }
            else if let d = number(t).map({ Int($0) }) {
                r.dynamicRange = [100: .dr100, 200: .dr200, 400: .dr400][d]
            }
        }
        // GRAIN
        if let l = sections["grain"] {
            let t = norm(l.joined(separator: " "))
            if t.contains("OFF") { r.grain = t.contains("LARGE") ? .offLarge : .off }
            else {
                let strong = t.contains("STRONG"), large = t.contains("LARGE")
                r.grain = strong ? (large ? .strongLarge : .strongSmall) : (large ? .weakLarge : .weakSmall)
            }
        }
        // COLOR CHROME
        func chrome(_ ls: [String]) -> ColorChrome {
            let t = norm(ls.joined(separator: " "))
            if t.contains("STRONG") { return .strong }
            if t.contains("WEAK") { return .weak }
            return .off
        }
        if let l = sections["cc_effect"] { r.colorChromeEffect = chrome(l) }
        if let l = sections["cc_blue"] { r.colorChromeBlue = chrome(l) }

        // WB
        if let l = sections["wb"] {
            let t = norm(l.joined(separator: " "))
            if let kStr = firstMatch("(\\d{4,5})\\s*K", t, group: 1), let k = Int(kStr) {
                r.wbKelvin = k; r.whiteBalance = .colorTemp
            } else {
                for (alias, mode) in wbModes where t.contains(alias) { r.whiteBalance = mode; break }
            }
            if let rs = firstMatch("([+-]?\\d+)\\s*RED", t, group: 1) ?? firstMatch("RED\\s*([+-]?\\d+)", t, group: 1) {
                r.wbShiftRed = Int(rs)
            }
            if let bs = firstMatch("([+-]?\\d+)\\s*BLUE", t, group: 1) ?? firstMatch("BLUE\\s*([+-]?\\d+)", t, group: 1) {
                r.wbShiftBlue = Int(bs)
            }
        }

        // tones / color / sharp / clarity
        if let l = sections["highlight"] { r.highlightTone = number(l.joined(separator: " ")) }
        if let l = sections["shadow"] { r.shadowTone = number(l.joined(separator: " ")) }
        if let l = sections["color"], let v = number(l.joined(separator: " ")) { r.color = Int(v) }
        if let l = sections["sharpness"], let v = number(l.joined(separator: " ")) { r.sharpness = Int(v) }
        if let l = sections["clarity"], let v = number(l.joined(separator: " ")) { r.clarity = Int(v) }

        // NR
        if let l = sections["nr"], let v = number(l.joined(separator: " ")) {
            r.noiseReduction = nrMap[Int(v)]
        }

        // notes
        if let l = sections["iso"] { r.notes["iso"] = l.joined(separator: " ") }
        if let l = sections["exp_comp"] { r.notes["exposure_compensation"] = l.joined(separator: " ") }

        return r
    }
}
